tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+    @field
+    def name(self) -> str:
+        return "parent replaced product name"
+
+
+@handle_urls(URL, to_return=SubclassReplacedProduct)
+class SubclassReplacedProductPage(ParentReplacedProductPage):
+    @field
+    def name(self) -> str:
+        return "subclass replaced product name"
+
+
+@inlineCallbacks
+def test_item_to_return_in_handle_urls_subclass(caplog) -> None:
+    """Same case as with the ``test_item_to_return_in_handle_urls()`` case above
+    but the ``to_return`` is declared in the subclass.
+    """
+    item, deps = yield crawl_item_and_deps(SubclassReplacedProduct)
+    assert "UndeclaredProvidedTypeError:" in caplog.text
+    assert item is None
+    assert_deps(deps, {}, size=0)
+
+    # Requesting the underlying item class from the parent PO should still work.
+    item, deps = yield crawl_item_and_deps(ParentReplacedProduct)
+    assert item == ParentReplacedProduct(name="parent replaced product name")
+    assert_deps(deps, {"item": ParentReplacedProduct})
+
+    # calling the actual page objects should still work
+
+    item, deps = yield crawl_item_and_deps(ParentReplacedProductPage)
+    assert item == ParentReplacedProduct(name="parent replaced product name")
+    assert_deps(deps, {"page": ParentReplacedProductPage})
+
+    item, deps = yield crawl_item_and_deps(SubclassReplacedProductPage)
+    assert item == ParentReplacedProduct(name="subclass replaced product name")
+    assert_deps(deps, {"page": SubclassReplacedProductPage})
+
+
+@attrs.define
+class StandaloneProduct:
+    name: str
+
+
+@handle_urls(URL, to_return=StandaloneProduct)
+class StandaloneProductPage(ItemPage):
+    @field
+    def name(self) -> str:
+        return "standalone product name"
+
+
+@inlineCallbacks
+def test_item_to_return_standalone(caplog) -> None:
+    """Same case as with ``test_item_to_return_in_handle_urls()`` above but the
+    page object doesn't inherit from something like ``ItemPage[ItemClass]``
+    """
+    item, deps = yield crawl_item_and_deps(StandaloneProduct)
+    assert "UndeclaredProvidedTypeError:" in caplog.text
+    assert item is None
+    assert_deps(deps, {}, size=0)
+
+    # calling the actual page object should still work
+    item, deps = yield crawl_item_and_deps(StandaloneProductPage)
+    assert item == {"name": "standalone product name"}
+    assert_deps(deps, {"page": StandaloneProductPage})
+
+
+@attrs.define
+class Morty:
+    name: str
+
+
+@handle_urls(URL)
+class RickSanchezPage(ItemPage[Morty]):
+    @field
+    def name(self) -> str:
+        return "from basic rick"
+
+
+@handle_urls(URL, instead_of=RickSanchezPage)
+class RickSanchezC137Page(ItemPage[Morty]):
+    @field
+    def name(self) -> str:
+        return "wubba lubba dub dub"
+
+
+@inlineCallbacks
+def test_item_return_with_overrides() -> None:
+    item, deps = yield crawl_item_and_deps(Morty)
+    assert item == Morty(name="wubba lubba dub dub")
+    assert_deps(deps, {"item": RickSanchezC137Page})
```

Comment:
I assume the "wubba lubba dub dub" Morty is returned because of an override, not because of a priority clash (both pages are registered for Morty item). Would it make sense to check that it's the case, by adding priority to the example, or adding another test with priorities?

Reply:
I've re-written these tests to support 3 different variations. https://github.com/scrapinghub/scrapy-poet/pull/88/commits/e184c6f09d210cdd84b1ae2abd5c9e904d0b4ce3